hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/streaming/BlockState.java

```diff
@@ -2,30 +2,152 @@
 import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.hiero.block.api.BlockItemSet;
 import org.hiero.block.api.PublishStreamRequest;
 
 /**
- * Track the current block state
+ * Represents the state of a block being streamed to block nodes. This class maintains the block items,
+ * completion status, and request generation for a specific block number.
+ * The block state goes through the following lifecycle:
+ * <ul>
+ *     <li>Created when a new block is opened</li>
+ *     <li>Block items are added sequentially</li>
+ *     <li>Requests are generated from accumulated items capped by a configurable batch size</li>
+ *     <li>Block is marked as complete after all items including BlockProof are added</li>
+ * </ul>
  */
 public class BlockState {
+    private static final Logger logger = LogManager.getLogger(BlockState.class);
+
+    /**
+     * Enum representing the state of a block item.
+     */
+    enum ItemState {
+        /**
+         * The item hasn't been encountered yet.
+         */
+        NIL,
+        /**
+         * The item has been added to the pending items queue, but not yet included in a request.
+         */
+        ADDED,
+        /**
+         * The item has been included - or packed - into a request.
+         */
+        PACKED,
+        /**
+         * The item has been sent to a block node.
+         */
+        SENT
+    }
+
+    /**
+     * Simple record for tracking request information, in particular if a given request has been sent to a block node.
+     *
+     * @param index the index (starting with 0) of the request
+     * @param request the actual request object that gets sent to the block node
+     * @param isSent flag indicating if this request has been sent to a block node
+     */
+    record RequestWrapper(int index, PublishStreamRequest request, AtomicBoolean isSent) {}
+
+    /**
+     * Record for capturing information specific to a item, such as what request it is associated with and what state
+     * the item is in.
+     *
+     * @param state the current state of the item (e.g. not seen vs sent)
+     * @param requestIndex the index of the request the item is associated with, else -1 if it hasn't been assigned to
+     *                     a request yet
+     */
+    record ItemInfo(AtomicReference<ItemState> state, AtomicInteger requestIndex) {
+        ItemInfo() {
+            this(new AtomicReference<>(ItemState.NIL), new AtomicInteger(-1));
+        }
+
+        /**
+         * Marks this item as added to the block state.
+         */
+        boolean addedInBlockState() {
```

Comment:
‼️ This name reads like a POJO getter that returns, "If the item was added to the `BlockState` when the method was called."

But it _actually_,
 - May have a side-effect to mark the item as added; and,
 - Returns `true` only if the item was **NOT** added at the start of the method. 😳 

This is wildly counter-intuitive and would be easy to accomplish using more unsurprising idioms.

Reply:
Discussed out-of-band, @timfn-hg will address in future PRs.